server/datastore/mysql/operating_systems.go

```diff
@@ -124,3 +124,40 @@ func getHostOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostID ui
 	}
 	return &os, nil
 }
+
+func (ds *Datastore) CleanupHostOperatingSystems(ctx context.Context) error {
+	// delete host_operating_system records associated with any non-existent hosts (e.g., expired hosts)
+	stmt := `
+	DELETE FROM host_operating_system
```

Comment:
It will be way more efficient to do this when the host is actually deleted in https://github.com/fleetdm/fleet/blob/main/server/datastore/mysql/hosts.go#L294

Reply:
Indeed. I commented the same [here](https://github.com/fleetdm/fleet/pull/7154#discussion_r943020273).

Once you do that, then this first `DELETE` in `CleanupHostOperatingSystems` will not be necessary. (Expired hosts cleanup and explicit host deletion both use `ds.DeleteHost` which deletes all related tables in a tx.)

Reply:
Thanks, I didn't know we had this!